cpp/src/parquet/column_reader.cc

```diff
@@ -353,6 +372,37 @@ std::shared_ptr<Page> SerializedPageReader::NextPage() {
   return std::shared_ptr<Page>(nullptr);
 }
 
+std::shared_ptr<Buffer> SerializedPageReader::DecompressPage(int compressed_len,
+                                                             int uncompressed_len,
+                                                             const uint8_t* page_buffer) {
+  // Grow the uncompressed buffer if we need to.
+  if (uncompressed_len > static_cast<int>(decompression_buffer_->size())) {
+    PARQUET_THROW_NOT_OK(decompression_buffer_->Resize(uncompressed_len, false));
+  }
+
+  if (current_page_header_.type != format::PageType::DATA_PAGE_V2) {
+    PARQUET_THROW_NOT_OK(
+        decompressor_->Decompress(compressed_len, page_buffer, uncompressed_len,
+                                  decompression_buffer_->mutable_data()));
+  } else {
+    // The levels are not compressed in V2 format
+    const auto& header = current_page_header_.data_page_header_v2;
+    int32_t levels_length =
+        header.repetition_levels_byte_length + header.definition_levels_byte_length;
+    uint8_t* decompressed = decompression_buffer_->mutable_data();
+    memcpy(decompressed, page_buffer, levels_length);
```

Comment:
Do we know for sure that `uncompressed_len` includes the size of the uncompressed encoded levels?

Reply:
That was my interpretation of this comment in [parquet.thrift](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L622) :
```
  /** Uncompressed page size in bytes (not including this header) **/
  2: required i32 uncompressed_page_size
```
I think the parquet-mr project seems to also follow this interpretation based on my superficial read of [this source](https://github.com/apache/parquet-mr/blob/806037c080dc477798d157cd4a54a81240a85d37/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java#L1248).

Happy to ask about this on the dev@parquet list to be sure.

Reply:
Sounds like we're in the clear here